services/powersupply.c

```diff
@@ -0,0 +1,67 @@
+#include "jd_services.h"
+#include "interfaces/jd_pins.h"
+#include "jacdac/dist/c/powersupply.h"
+
+struct srv_state {
+    SRV_COMMON;
+    uint8_t enabled;
+    int32_t output_voltage;
+    int32_t minimum_voltage;
+    int32_t maximum_voltage;
+    power_supply_params_t params;
+};
+
+REG_DEFINITION(                            //
+    psu_regs,                             //
+    REG_SRV_COMMON,                        //
+    REG_U8(JD_POWER_SUPPLY_REG_ENABLED), //
+    REG_I32(JD_POWER_SUPPLY_REG_OUTPUT_VOLTAGE),  //
+    REG_I32(JD_POWER_SUPPLY_REG_MINIMUM_VOLTAGE),  //
+    REG_I32(JD_POWER_SUPPLY_REG_MAXIMUM_VOLTAGE),  //
+)
+
+static void update_wiper_value(srv_t *state) {
+    int32_t value = state->output_voltage;
+
+    // should be handled client side, but check here too...
+    if (value < state->minimum_voltage || value > state->maximum_voltage)
+        return;
+
+    // Lazy implementation, use only the upper 22 bits.
+    value >>= 10;
+    state->params.potentiometer->set_wiper(state->params.wiper_channel, state->params.voltage_to_wiper(value));
+}
+
+static void reflect_register_state(srv_t *state) {
+    if (state->enabled) {
+        pin_set(state->params.enable_pin, (state->params.enable_active_lo) ? 0 : 1);
+        update_wiper_value(state);
+    }
+    else
+        pin_set(state->params.enable_pin, (state->params.enable_active_lo) ? 1 : 0);
+}
+
+void psu_process(srv_t *state) {
+    (void)state;
+}
+
+void psu_handle_packet(srv_t *state, jd_packet_t *pkt) {
+    service_handle_register_final(state, pkt, psu_regs);
```

Comment:
I think you want to see if return value is JD_POWER_SUPPLY_REG_OUTPUT_VOLTAGE (meaning that register changed) and if so sanitize the value of the register (clamp it to min/max); or you could always sanitize it in reflect_register_State